### PR TITLE
Adds primitive implementations of DISTINCT

### PR DESCRIPTION
**Pull request by systay**
The general purpose DISTINCT implementation uses a
lot of objects to make the output distinct.
This implementation avoids objects and uses primitive
sets to do work, which should be create less memory
pressure and is easier for the CPU to do efficiently.

### CODE REVIEW

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala [hunk 1]
```diff
@@ -345,6 +343,67 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
       .asInstanceOf[Predicate]
   }
 
+  private def distinctPipe(projections: Map[String, AstExpression],
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala]: Could we call this argument `groupingExpressions` like in the calling code?

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala [hunk 2]
```diff
@@ -345,6 +343,67 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
...
+        case v: NodeFromSlot =>
+          AllPrimitive(offsets :+ v.offset)
+        case v: RelationshipFromSlot =>
+          AllPrimitive(offsets :+ v.offset)
+        case _ =>
+          References
+      }
+    }
+
+    object Empty extends DistinctPhysicalOp {
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala]: `val Empty = AllPrimitice(Nil)`?

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        while (input.nonEmpty) { // Let's pull data until we find something not already seen
+          val next = input.next()
+          val keys: Array[Long] = primitiveSlots.map(next.getLongAt)
+          if (!seen.contains(keys)) {
+            seen.add(keys)
+            // Found something! Set it as the next element to yield, and exit
+            val outgoing = SlottedExecutionContext(slots)
+            setValuesInOutput.foreach {
+              _ (next, state, outgoing)
+            }
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala]:
```
for ( setter <- setValuesInOutput )
  setter(next, state, outgoing)
```

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/LongArraySet.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime.slotted.pipes
+
+import java.lang.Integer.highestOneBit
+import java.util
+
+class LongArraySet(capacity: Int = 32, longsPerEntry: Int) {
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/LongArraySet.scala]: I like the idea behind this set, but have some concerns:

 * A new general purpose primitive set should not live here, but along it's friends in `org.neo4j.collection.primitive.*`
 * To motivate this new primitive set, are there read/write benchmarks?
 * More testing of corner cases. `longsPerEntry == 0`? ScalaCheck would be perfect, if it wasn't for that fact that I'm suggesting to move this to java-land...
 * A small javadoc to describe the purpose of the set.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+      pullNextElementFromSource()
+
+      private def pullNextElementFromSource(): Unit = {
+        buffer = null
+        while (input.nonEmpty) { // Let's pull data until we find something not already seen
+          val next = input.next()
+          val keys: Array[Long] = primitiveSlots.map(next.getLongAt)
+          if (!seen.contains(keys)) {
+            seen.add(keys)
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala]: Instead of `contains` followed by `add`, `add` could return a boolean to describe whether the value was added or already there. That would half the number of `hashCode` computations.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedSinglePrimitivePipe.scala [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  expression.registerOwningPipe(this)
+
+  //===========================================================================
+  // Runtime code
+  //===========================================================================
+  protected def internalCreateResults(input: Iterator[ExecutionContext],
+                                      state: QueryState): Iterator[ExecutionContext] = {
+
+
+    new Iterator[ExecutionContext] {
```
Comment (fickludd) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedSinglePrimitivePipe.scala]: This is a `PrefetchingIterator`. Maybe we could create a scala version of that to inherit from, so we don't have to implement it in scala more than once? E.g.

```
abstract class PrefetchingIterator[T] extends Iterator[T] {
  private var _next: T = _
  def fetchNext(): T
  def hasNext: Boolean = _next != null
  def next(): T =
    if (isEmpty)
      Iterator.empty.next() // Throw a good exception
    else {
      val toReturn = _next
      fetchNext()
      toReturn
    }
}
```

**Review by fickludd — CHANGES_REQUESTED** [review]
These specializations should help out a lot I think, nice find + fix. Before merging this though, I think the new primitive collection should move to live with the other primitive collections, and get some more love. Also: I suspect we could need some more tests somewhere to cover all these 3 implementations of distinct.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala (hunk 1 above)
Comment (systay) [reply enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala]: It's not actually grouping, so I rather change the calling code... But OK, consistency is king :)

**Comment by systay** [comment]
Waiting for a 3.5 branch

**Comment by fickludd** [comment]
@systay I'm happy with the implementation now!

If you could just reassure me that there is existing test coverage, and maybe share some numbers on the observed improvement, I'm happy to merge :).

File: community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/PrefetchingIterator.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime
+
+abstract class PrefetchingIterator[T] extends Iterator[T] {
```
Comment (henriknyman) [inline community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/PrefetchingIterator.scala:22]: Can't we use `org.neo4j.helpers.collection.PrefetchingIterator` instead to avoid having duplicate functionality?

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            for (setter <- setValuesInOutput) {
+              setter(next, state, outgoing)
+            }
+
+            return Some(outgoing)
+          }
+        }
+
+        None
+      }
```
Comment (henriknyman) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala:79]: If you have measurement indicating that this way of _iterating_ is faster than the old way in `DistinctSlottedPipe` it would be good to apply it to that too.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedSinglePrimitivePipe.scala [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+          val next = input.next()
+          val id = next.getLongAt(offset)
+          if (seen.add(id)) {
+            // Found something! Set it as the next element to yield, and exit
+            val outgoing = SlottedExecutionContext(slots)
+            val outputValue = expression(next, state)
+            setInSlot(outgoing, outputValue)
+            return Some(outgoing)
+          }
+        }
```
Comment (henriknyman) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedSinglePrimitivePipe.scala:68]: If the runtime cost is negligible it would be nice if these pipes could share an abstract base class to avoid duplicating the control flow code.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+
+  private def buildKey(next: ExecutionContext): Array[Long] = {
+    val keys = new Array[Long](primitiveSlots.length)
+    var i = 0
+    while (i < primitiveSlots.length) {
+      keys(i) = next.getLongAt(primitiveSlots(i))
+      i += 1
+    }
```
Comment (henriknyman) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala:89]: There is a special case of this pipe where the slots are consecutive in the execution context when we could use System.arraycopy here, but I don't know if it is worth optimizing for.

File: community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/LongArraySet.scala [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  * This set will keep all it's state in a single long[] array, marking unused slots
+  * using 0xF000000000000000L, a value that should never be used for node or relationship id's.
+  *
+  * The set will be resized when either probing has to go on for too long when doing inserts,
+  * or the load factor reaches 75%.
+  *
+  * @param capacity The initial capacity for the set. Must be a power of 2
+  * @param longsPerEntry All arrays in the set must be of this length
+  */
+class LongArraySet(capacity: Int = 32, longsPerEntry: Int) {
```
Comment (henriknyman) [inline community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/LongArraySet.scala]: It would be good to have better understanding of the performance characteristics of this implementation over a balanced range of data sets...

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala [hunk 12]
```diff
@@ -299,52 +274,12 @@ class SlottedPipeBuilder(fallback: PipeBuilder,
...
-  }
-
-  private def buildPredicate(expr: frontEndAst.Expression)
-                            (implicit context: PipeExecutionBuilderContext, planContext: PlanContext): Predicate = {
-    val rewrittenExpr: frontEndAst.Expression = rewriteAstExpression(expr)
-
-    expressionConverters.toCommandPredicate(rewrittenExpr).rewrite(KeyTokenResolver.resolveExpressions(_, planContext))
-      .asInstanceOf[Predicate]
-  }
-
```
Comment (henriknyman) [inline enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala:347]: Why do we need to move this?

**Review by henriknyman — COMMENTED** [review]
Still awaiting benchmark results to see what the gains are

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala (hunk 12 above)
Comment (systay) [reply enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/SlottedPipeBuilder.scala:347]: I moved all private methods to the bottom.

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala (hunk 8 above)
Comment (systay) [reply enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala:79]: Hmm... Thanks for pointing this out. Not sure I have any reasons to prefer this... Looks like the 
builtin Scala collection `filter` does something very similar... I'll double check

File: community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/PrefetchingIterator.scala (hunk 7 above)
Comment (systay) [reply community/cypher/runtime-util/src/main/scala/org/neo4j/cypher/internal/runtime/PrefetchingIterator.scala:22]: That is a Java Iterator, and this is a Scala one...

File: enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala (hunk 8 above)
Comment (systay) [reply enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala:79]: I understand why I did this now. In the old `DistinctSlottedPipe`, we build the output execution context and execute the expressions on that context. Here, the "distinctification" is done before creating the output `ExecutionContext`. I don't think this pattern would be applicable for the old pipe.
